File: go/src/impl/contract.go
package impl

import (
	"errors"
	"github.com/redstone-finance/redstone-contracts-wasm/go/src/common_types"
	"github.com/redstone-finance/redstone-contracts-wasm/go/src/types"
)

type PstContract struct {
	state types.PstState
}

// Handle the function that contract developers actually need to implement
func (c *PstContract) Handle(action common_types.Action, actionBytes []byte) (interface{}, common_types.ActionResult, error) {
	fn := action.Function

	// this is how you can access functions imported from js
	/*
		console.Log("Block height", block.Height())
		console.Log("Block indep_hash", block.IndepHash())
		console.Log("Block timestamp", block.Timestamp())

		console.Log("Transaction id", transaction.Id())
		console.Log("Transaction owner", transaction.Owner())
		console.Log("Transaction target", transaction.Target())

		console.Log("Contract id", contract.Id())
		console.Log("Contract owner", contract.Owner())
	*/

	clonedState := c.CloneState().(types.PstState)

	switch fn {
	case "transfer":
		var transfer types.TransferAction
		err := transfer.UnmarshalJSON(actionBytes)
		if err != nil {
			return nil, nil, err
		}
		state, err := Transfer(clonedState, transfer)
		return state, nil, err
	case "evolve":
		var evolve types.EvolveAction
		err := evolve.UnmarshalJSON(actionBytes)
		if err != nil {
			return nil, nil, err
		}
		state, err := Evolve(clonedState, evolve)
		return state, nil, err
	case "balance":
		var balance types.BalanceAction
		err := balance.UnmarshalJSON(actionBytes)
		if err != nil {
			return nil, nil, err
		}
		result, err := Balance(clonedState, balance)
		return nil, result, err
	case "foreignCall":
		var foreignCall types.ForeignCallAction
		err := foreignCall.UnmarshalJSON(actionBytes)
		if err != nil {
			return nil, nil, err
		}
		state, err := ForeignCall(clonedState, foreignCall)
		return state, nil, err
	default:
		return nil, nil, errors.New("[RE:WTF] unknown function: " + fn)
	}
}

func (c *PstContract) InitState(stateJson string) {
	var state types.PstState
	err := state.UnmarshalJSON([]byte(stateJson))
	if err != nil {
		return // TODO: throw in a similar way as in handle
	}
	c.UpdateState(&state)
}

func (c *PstContract) UpdateState(newState interface{}) {
	// note: we're first type asserting here to the pointer to types.PstState
	// - and then retrieving value from the pointer
	c.state = *(newState.(*types.PstState))
}

func (c *PstContract) CurrentState() interface{} {
	return c.state
}

func (c *PstContract) CloneState() interface{} {
	json, _ := c.state.MarshalJSON()
	state := types.PstState{}
	err := state.UnmarshalJSON(json)
	if err != nil {
		// TODO: return error
		return types.PstState{}
	}

	return state
}

File: rust/pst/implementation/src/contract.rs
use async_recursion::async_recursion;

use warp_pst::action::{Action, ActionResult};
use warp_pst::state::State;

use crate::actions::{Actionable, *};
use warp_wasm_utils::contract_utils::js_imports::{Block, Contract, log, SmartWeave, Transaction};

#[async_recursion(?Send)]
pub async fn handle(state: State, action: Action) -> ActionResult {
    let original_caller = Transaction::owner();
    let effective_caller = SmartWeave::caller();

    //Example of accessing functions imported from js:
    // log("log from contract");
    // log(&("Transaction::id()".to_owned() + &Transaction::id()));
    // log(&("Transaction::owner()".to_owned() + &Transaction::owner()));
    // log(&("Transaction::target()".to_owned() + &Transaction::target()));

    // log(&("Block::height()".to_owned() + &Block::height().to_string()));
    // log(&("Block::indep_hash()".to_owned() + &Block::indep_hash()));
    // log(&("Block::timestamp()".to_owned() + &Block::timestamp().to_string()));

    // log(&("Contract::id()".to_owned() + &Contract::id()));
    // log(&("Contract::owner()".to_owned() + &Contract::owner()));

    // log(&("SmartWeave::caller()".to_owned() + &SmartWeave::caller()));

    // for vrf-compatible interactions
    /*log(&("Vrf::value()".to_owned() + &Vrf::value()));
    log(&("Vrf::randomInt()".to_owned() + &Vrf::randomInt(7).to_string()));*/

    match action {
        Action::Transfer(action) => action.action(effective_caller, state),
        Action::Balance(action) => action.action(effective_caller, state),
        Action::Evolve(action) => action.action(effective_caller, state),
        Action::ForeignRead(action) => action.action(effective_caller, state).await,
        Action::ForeignWrite(action) => action.action(effective_caller, state).await,
    }
}
